op-challenger/game/keccak/merkletree.go

```diff
@@ -0,0 +1,98 @@
+package keccak
+
+import (
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// KeccakTreeDepth is the depth of the keccak tree.
+const KeccakTreeDepth = 16
+
+// MaxLeafCount is the maximum number of leaves in the keccak tree.
+var MaxLeafCount = 1<<KeccakTreeDepth - 1 // 2^16 - 1
+
+type merkleNode struct {
+	data common.Hash
+
+	left  *merkleNode
+	right *merkleNode
+}
+
+// MerkleProof is a keccak merkle proof, which is a list of hashes that prove the inclusion of a leaf in the merkle tree.
+type MerkleProof [][]byte
+
+// toSized converts a [MerkleProof] to a list of commitment hashes.
+func ToSized(m MerkleProof) []common.Hash {
+	var sized []common.Hash
+	for _, proof := range m {
+		// SAFETY: if the proof is less than 32 bytes, it will be padded with 0s
+		if len(proof) < 32 {
+			proof = append(proof, make([]byte, 32-len(proof))...)
+		}
+		// SAFETY: the proof is 32 or more bytes here, so it will be truncated to 32 bytes
+		sized = append(sized, common.Hash(proof[:32]))
+	}
+	return sized
+}
+
+type Hashable interface {
+	Hash() common.Hash
+}
+
+type MerkleTree struct {
+	Root *merkleNode
+
+	// The number of leaves in the tree
+	LeafCount int
+}
+
+func NewMerkleTree(leafCount int) *MerkleTree {
+	return &MerkleTree{
+		Root:      &merkleNode{data: common.Hash{}},
+		LeafCount: leafCount,
+	}
+}
+
+func (m *MerkleTree) AddLeaf(leaf types.Leaf) {
+	// First hash the leaf data
```

Comment:
Actually this is probably bad advice.  The geth trie implementation will be a hexary tree and the deposit contract uses a binary tree. Given our contract merkle tree implementation came from the deposit contract it will be binary as well, so we can't just use the geth Trie class.

I would be surprised if we don't have a binary merkle tree implementation somewhere in our codebase already though.

Reply:
I don't believe we do, and from what i've explored, other implementations out there are not easy to work with for our use case. I sured up the implementation and added a bunch of tests. These only go so far as to assert the expected output, but isn't hooked up to the contract yet. I verified the outputs match what @clabby implemented in https://github.com/ethereum-optimism/optimism/pull/9147.

If we are able to update https://github.com/ethereum-optimism/optimism/pull/9147 to work with the merkle tree in this pr, we can then check against the contracts I believe.